### PR TITLE
Enable Fleet Server

### PR DESCRIPTION
**Pull request by mtojek**
This PR adjusts the Elastic stack created by the elastic-package to supports Fleet Server.

Issue: https://github.com/elastic/elastic-package/issues/278

### CODE REVIEW

**Comment by mtojek** [comment]
Right now I can't enroll the agent to any policy:

```
➜  elastic-package git:(278-support-fleet-server) ✗ docker logs 47b5c520a12f -f
Performing setup of Fleet in Kibana
The Elastic Agent is currently in BETA and should not be used in production

2021-03-08T14:53:43.418Z	INFO	application/enroll_cmd.go:287	Generating self-signed certificate for Fleet Server
2021-03-08T14:53:43.787Z	INFO	application/enroll_cmd.go:435	Spawning Elastic Agent daemon as a subprocess to complete bootstrap process.
2021-03-08T14:53:44.793Z	INFO	application/enroll_cmd.go:513	waiting for Elastic Agent to start Fleet Server
2021-03-08T14:53:45.795Z	INFO	application/enroll_cmd.go:531	Fleet Server - Starting
2021-03-08T14:53:47.428Z	WARN	application/enroll_cmd.go:331	Remote server is not ready to accept connections, will retry in a moment.
2021-03-08T14:54:47.360Z	INFO	application/enroll_cmd.go:338	Retrying to enroll...
2021-03-08T14:54:47.363Z	WARN	application/enroll_cmd.go:331	Remote server is not ready to accept connections, will retry in a moment.
```

**Comment by nchaulet** [comment]
@mtojek I was able to have it running on 8.0 by adding `- "FLEET_SERVER_INSECURE_HTTP=1"`

**Comment by mtojek** [comment]
@nchaulet I added the suggested env and now I'm getting this error from Agent:

```
2021-03-09T18:37:09.755Z	ERROR	application/fleet_gateway.go:185	failed to dispatch actions, error: fail to generate program configuration: expecting Dict and received *transpiler.Key for '0'
2021-03-09T18:37:09.755Z	WARN	status/reporter.go:233	Elastic Agent status changed to: 'degraded'
2021-03-09T18:37:09.755Z	INFO	status/reporter.go:233	Elastic Agent status changed to: 'online'
```

Package revision in the agent's structure is "null".

File: internal/install/static_snapshot_yml.go
Comment (mtojek) [inline internal/install/static_snapshot_yml.go:94]: @blakerouse @ruflin @nchaulet I watched the observability demo session and @blakerouse's presentation (good job!) about running the agent with fleet server in a container and I'm confused about available configuration options. Recently I've removed FLEET_URL and KIBANA_HOST from global envs, but I saw @blakerouse used them. Unfortunately with current blockers it's hard to determine the correct set.

Shall I ask you to review both `kibana.config.yml` and `snapshot.yml` and recommend the best configuration for 7.13?

File: internal/install/static_kibana_config_yml.go
Comment (ruflin) [inline internal/install/static_kibana_config_yml.go:21]: This will change as part of https://github.com/elastic/beats/pull/24713 and https://github.com/elastic/kibana/pull/94364

Comment (mtojek) [reply internal/install/static_kibana_config_yml.go:21]: Thank you for reviewing this. What is the recommendation though? Should I wait until these PRs are merged?

File: internal/install/static_snapshot_yml.go
Comment (blakerouse) [reply internal/install/static_snapshot_yml.go:94]: @mtojek `FLEET_URL` should not be required, but at the moment there is an issue that requires it. I am working to solve this issue. As for the `KIBANA_HOST` that is required if you want the container to perform the setup of Kibana working with Fleet, which when running with `FLEET_SERVER_ENABLED` is required, unless you are running at some hook on the Kibana container instead.

Comment (mtojek) [reply internal/install/static_snapshot_yml.go:94]: > FLEET_URL should not be required, but at the moment there is an issue that requires it. I am working to solve this issue.

Would you mind linking this issue for tracking purposes?

> As for the KIBANA_HOST that is required if you want the container to perform the setup of Kibana working with Fleet, which when running with FLEET_SERVER_ENABLED is required, unless you are running at some hook on the Kibana container instead.

In this setup we've a long running Elastic cluster (Elasticsearch, Kibana, Package Registry) and two agents (so far):
1. Separate agent's container which is used for system tests (reassigning policies).
2. Agent's container in the Kubernetes cluster.

which gives us two agent containers, each one with own FleetServer inside. Can Kibana handle it without any problems?

File: internal/install/static_snapshot_yml.go
Comment (ruflin) [inline internal/install/static_snapshot_yml.go:93]: You need to add `FLEET_SERVER_HOST=0.0.0.0`. This will allow an Elastic Agent running on your machine to enroll into this fleet-server.

Comment (mtojek) [reply internal/install/static_snapshot_yml.go:92]: Ok, added this one too.

File: internal/install/static_snapshot_yml.go
Comment (ruflin) [reply internal/install/static_snapshot_yml.go:93]: Two fleet-server should just work. If not, please let us know.

File: internal/install/static_kibana_config_yml.go
Comment (ruflin) [reply internal/install/static_kibana_config_yml.go:21]: I would like to switch over elastic-package to fleet-server as soon as possible to have early testing. At the same time I do not want to impact the integrations team with potential bugs / changes. The above changes can only land if a dependency from endpoint is also merged at the same time. I think an easy trick on our side will be to just ahve both config options in already so things will keep working. @nchaulet is this assumption correct?

**Comment by mtojek** [comment]
Status update:

I had a 1-1 with @ruflin regarding the configuration. I learnt much about the Fleet Server package, which seems to be the missing part for the system test runner to handle. I will adjust its implementation.

I will also use environment variables specified in the: https://github.com/elastic/beats/blob/master/x-pack/elastic-agent/pkg/agent/cmd/container.go#L61

File: internal/install/static_kibana_config_yml.go
Comment (nchaulet) [reply internal/install/static_kibana_config_yml.go:21]: Yes thats correct we can have both options as soon my PR for fleet server hosts is merged (hopefully soon :) ) https://github.com/elastic/kibana/pull/94364

File: internal/install/static_snapshot_yml.go
Comment (ruflin) [inline internal/install/static_snapshot_yml.go:93]: How does the fleet-server know username / password for Kibana?

File: internal/install/static_snapshot_yml.go
Comment (ruflin) [inline internal/install/static_snapshot_yml.go:111]: I wonder how in this scenario the elastic-agent will get the right enrollment token. We might still have to read it from Kibana.

File: internal/install/static_snapshot_yml.go
Comment (mtojek) [reply internal/install/static_snapshot_yml.go:93]: It's standard elastic/changeme

**Comment by mtojek** [comment]
Status update:

I increased the interval between healthchecks and it skips the problematic gap when the Fleet Server is not available, but the flakiness still persists.

It seems that you can't assign the policy to the agent. There is no error in logs, but the policy revision is null.

**Comment by mtojek** [comment]
jenkins run the tests please

**Comment by mtojek** [comment]
jenkins run the tests please

**Comment by mtojek** [comment]
Passed: https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Felastic-package/detail/PR-279/32/pipeline
Failed: https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Felastic-package/detail/PR-279/31/pipeline/

I suspect that there is some flakiness around, maybe related to healthchecks.

**Comment by mtojek** [comment]
It looks like the PR is ready for review. Would you mind looking at it one more time to see if I haven't missed anything?

File: internal/install/static_snapshot_yml.go
Comment (nchaulet) [reply internal/install/static_snapshot_yml.go:112]: It's probably reading it from ES but how does the agent know what ES to call?

File: internal/install/static_snapshot_yml.go
Comment (blakerouse) [inline internal/install/static_snapshot_yml.go:95]: You should not need this anymore. By default Elastic Agent will start Fleet Server with it bound to `0.0.0.0`.

File: internal/install/static_snapshot_yml.go
Comment (blakerouse) [inline internal/install/static_snapshot_yml.go:92]: I would rather see you run it without this flag. Why run it insecurely?

File: internal/install/static_snapshot_yml.go
Comment (blakerouse) [inline internal/install/static_snapshot_yml.go:86]: Can add `--insecure` to the `curl` command and change to `https` if you remove the `FLEET_SERVER_INSECURE_HTTP` below.

File: internal/install/static_snapshot_yml.go
Comment (blakerouse) [reply internal/install/static_snapshot_yml.go:112]: Would prefer the usage of `https` here as well. But you will still need the `FLEET_INSECURE=1` so it sets `ssl.verification_mode: none`.

**Review by ruflin — APPROVED** [review]
I did a quick test on 7.13 and all works as expected. I must confess the setup took a bit longer then I expected but this is much more to the setup on the fleet-server / Fleet end then this scripting and we need to make improvements there.

File: internal/install/static_snapshot_yml.go
Comment (blakerouse) [inline internal/install/static_snapshot_yml.go:105]: The `status` command landed, so it would be better to run that instead of this type of check.

`test: "./elastic-agent status"`

Should be enough, as it returns exit code 0 when the agent is healthy.

File: internal/install/static_snapshot_yml.go
Comment (mtojek) [reply internal/install/static_snapshot_yml.go:92]: For debugging purposes we can sniff network traffic and see requests/responses. It's not a production setup.

File: internal/install/static_snapshot_yml.go
Comment (mtojek) [reply internal/install/static_snapshot_yml.go]: Fixed

File: internal/install/static_snapshot_yml.go
Comment (mtojek) [reply internal/install/static_snapshot_yml.go]: Fixed

File: internal/install/static_snapshot_yml.go
Comment (mtojek) [reply internal/install/static_snapshot_yml.go:95]: Something doesn't work (connectivity issue), I will try to revert this one.

**Comment by ruflin** [comment]
@mtojek I would not block this PR on all the changes. We have a working version which can be used and we can always iterate on top of it. I think it is more important to start using fleet-server for all the testing instead of having all the perfect params.

**Pull request merged**
